### PR TITLE
[5044] Kl 2022 09 dashboard token export

### PR DESCRIPTION
**Pull request by fuzzylogic2000**
@phillimorland Could you have a look at the template? There is the table-stuff from the export in, but also a form, that is a bit out of place and a sentence about the number, which is not in the design. But let's leave the sentence and then ask if the wording is alright or were to actually put that number. But for the table and the form could you make the template nicer and make the generate button be on the left like in the design?

@Rineee I do create the tokens in a bit of a weird way because I didn't want to use the batch create as the docs say that that doesn't use the save from the model and that is what we want to use to make sure the token is unique. 
https://docs.djangoproject.com/en/4.1/ref/models/querysets/#bulk-create
What do you think?

### CODE REVIEW

**Comment by phillimorland** [comment]
@fuzzylogic2000 yes can totally take a look but may make more sense to do it after the BS is reviewed and merged as will change how I do it, just the fixing PR again :p

**Review by Rineee — APPROVED** [review]
Really cool, code looks great!

I did a test run with 100000 tokens, it took 15 min to generate them :flushed:  Dont know if we could somehow optimize it? Or will it be much quicker on the server?
Maybe we could disable the button and/or add a helptext that it can take a long time if a huge amount is generated?

**Comment by Rineee** [comment]
> @phillimorland Could you have a look at the template? There is the table-stuff from the export in, but also a form, that is a bit out of place and a sentence about the number, which is not in the design. But let's leave the sentence and then ask if the wording is alright or were to actually put that number. But for the table and the form could you make the template nicer and make the generate button be on the left like in the design?
> 
> @Rineee I do create the tokens in a bit of a weird way because I didn't want to use the batch create as the docs say that that doesn't use the save from the model and that is what we want to use to make sure the token is unique. https://docs.djangoproject.com/en/4.1/ref/models/querysets/#bulk-create What do you think?

True, we need the save method to be called.. or maybe we could change that to happen when generating the tokens? Hm, I dont know.. then it would still not happen when adding a token in django admin. I also could not really find information on how bulk_create works with unique constraints.. did you try out if it is much faster with bulk_create?

File: meinberlin/apps/votes/dashboard.py
Comment (fuzzylogic2000) [reply meinberlin/apps/votes/dashboard.py:13]: Ups, I think, I took it from the design and then changed the other places, but obviously not the headline, then. Good catch!

File: meinberlin/apps/votes/views.py
Comment (fuzzylogic2000) [reply meinberlin/apps/votes/views.py:44]: Yeah, but as it doesn't really do anything... Still, might look cleaner. Will give it a go.

**Comment by fuzzylogic2000** [comment]
@Rineee Yes, I think, I will just make second branch to try out the bulk stuff. Maybe it would be better to make the uuid field unique anyway and not unique together with the module and then using the save-method to make sure that that is the case. With changing that, it should work for the save and hopefully the bulk create, right? Or what was the reason again for the way we do it now?

**Comment by Rineee** [comment]
> @Rineee Yes, I think, I will just make second branch to try out the bulk stuff. Maybe it would be better to make the uuid field unique anyway and not unique together with the module and then using the save-method to make sure that that is the case. With changing that, it should work for the save and hopefully the bulk create, right? Or what was the reason again for the way we do it now?

Yes, good idea to try it out! 
I think we made it unique together with module, because we were fearing to run out of tokens at some point :sweat_smile: And the whole stuff in the save method is just there to make sure that really the amount of codes is generated, even if there was one that already existed.. if that is what you meant?

**Comment by fuzzylogic2000** [comment]
I moved the token creation to another branch now, so we can merge and try this if it is finished. I would like to fix the export for > 1,000,000 tokens first, though.

**Comment by fuzzylogic2000** [comment]
@Rineee I left the first commit without the splitting like the original commit, so that you can see what I changed and maybe more easily see what can be improved. Before merging, I would like to squash these commits though. (And I know that the wording is not yet like it is meant to be (for the sentence with how many codes there are and also for the export), but there are a few questions, I would like to talk about when it is on dev.)

File: meinberlin/apps/votes/views.py
Comment (Rineee) [inline meinberlin/apps/votes/views.py:28]: Is there a reason you dont use count()?

**Comment by Rineee** [comment]
> @Rineee I left the first commit without the splitting like the original commit, so that you can see what I changed and maybe more easily see what can be improved. Before merging, I would like to squash these commits though. (And I know that the wording is not yet like it is meant to be (for the sentence with how many codes there are and also for the export), but there are a few questions, I would like to talk about when it is on dev.)

@fuzzylogic2000 Really cool, looks good and works perfectly! So from my side good to merge after squash.

File: meinberlin/apps/votes/views.py
Comment (fuzzylogic2000) [reply meinberlin/apps/votes/views.py:28]: No, will do!

**Review by Rineee — APPROVED** [review]
Really nice!

**Pull request merged**
